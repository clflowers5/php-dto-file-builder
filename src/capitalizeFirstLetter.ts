function capitalizeFirstLetter(str: string): string {
  if (str.length > 1) {
    return str.charAt(0).toUpperCase() + str.slice(1);
  } else if (str.length > 0) {
    return str.charAt(0).toUpperCase();
  } else {
    return str;
  }
}

export default capitalizeFirstLetter;
