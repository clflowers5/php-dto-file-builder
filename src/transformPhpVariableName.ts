function transformPhpVariableName(variableName: string): string {
  if (variableName && variableName[0] !== '$') {
    return `$${variableName}`;
  } else {
    return variableName;
  }
}

export default transformPhpVariableName;
