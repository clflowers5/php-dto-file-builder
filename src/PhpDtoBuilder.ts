import transformPhpVariableName from './transformPhpVariableName';

interface DtoBuilder {
  addPrivateVariable(variableName: string): DtoBuilder;
  addPublicVariable(variableName: string): DtoBuilder;
  build(): string;
}

class PhpDtoBuilder implements DtoBuilder {
  private privateVariables: string[] = [];
  private publicVariables: string[] = [];

  constructor(private className: string) {}

  public addPrivateVariable(variableName: string): DtoBuilder {
    this.addNonEmptyStringToCollection(variableName, this.privateVariables);
    return this;
  }

  public addPublicVariable(variableName: string): DtoBuilder {
    this.addNonEmptyStringToCollection(variableName, this.publicVariables);
    return this;
  }

  public build(): string {
    return '';
  }

  private addNonEmptyStringToCollection(
    variableName: string,
    collection: string[]
  ): void {
    const trimmedVar = variableName.trim();
    if (trimmedVar) {
      collection.push(transformPhpVariableName(trimmedVar));
    }
  }
}

export default PhpDtoBuilder;
