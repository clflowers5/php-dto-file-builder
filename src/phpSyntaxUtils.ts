import transformPhpVariableName from './transformPhpVariableName';
import capitalizeFirstLetter from './capitalizeFirstLetter';

// todo: this could be refactored to a real class or something. idk.
function buildClass(className: string, privateVariables: string[]): string {
  return `<?php

class ${className} {
${buildPrivateVariableDeclarations(privateVariables)}
${buildSetters(privateVariables)}
${buildGetters(privateVariables, '' /* todo: here */)}
}
`;
}

function buildPrivateVariableDeclarations(privateVariables: string[]): string {
  return privateVariables.reduce((carry: string, item: string) => {
    carry += `${singleIndent()}private ${transformPhpVariableName(item)};\n`;
    return carry;
  }, '');
}

function buildGetters(variables: string[], types: string): string {
  return variables.reduce((carry: string, variableName: string) => {
    carry += `${singleIndent()}${buildGetter(variableName, '' /* todo: here */)}`;
    return carry;
  }, '');
}

function buildGetter(variableName: string, type: string): string {
  return `
    /**
     * @return ${type}
     */
    public function get${capitalizeFirstLetter(variableName)}() {
      return $this->${variableName};
    }
  `;
}

function buildSetters(variables: string[]): string {
  return variables.reduce((carry: string, variableName: string) => {
    carry += `${singleIndent()}${buildSetter(variableName, '' /* todo: here */)}`;
    return carry;
  }, '');
}

function buildSetter(variableName: string, type: string): string {
  return `
    /**
     * @return self
     */
    public function set${capitalizeFirstLetter(variableName)}(${transformPhpVariableName(variableName)}) {
      $this->${variableName} = ${transformPhpVariableName(variableName)};
      return $this;
    }
  `
}

/**
 * single indent = 4 spaces
 */
function singleIndent(): string {
  return '    ';
}

export { buildClass };
