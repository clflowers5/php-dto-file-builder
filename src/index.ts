import PhpDtoBuilder from './PhpDtoBuilder';

export { PhpDtoBuilder };
