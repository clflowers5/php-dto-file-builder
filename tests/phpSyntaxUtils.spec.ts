import { buildClass } from './../src/phpSyntaxUtils';

describe('buildClass', () => {
  it('snapshot variant 1', () => {
    const result = buildClass('MyClassHomeboy', [
      'one',
      'two',
      'three',
      'four',
      'five'
    ]);
    expect(result).toMatchSnapshot();
  });

  // it('snapshot variant 2', () => {

  // });
});
