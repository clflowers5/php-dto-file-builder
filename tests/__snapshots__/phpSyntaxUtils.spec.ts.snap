// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildClass snapshot variant 1 1`] = `
"<?php

class MyClassHomeboy {
    private $one;
    private $two;
    private $three;
    private $four;
    private $five;

    
    /**
     * @return self
     */
    public function setOne($one) {
      $this->one = $one;
      return $this;
    }
      
    /**
     * @return self
     */
    public function setTwo($two) {
      $this->two = $two;
      return $this;
    }
      
    /**
     * @return self
     */
    public function setThree($three) {
      $this->three = $three;
      return $this;
    }
      
    /**
     * @return self
     */
    public function setFour($four) {
      $this->four = $four;
      return $this;
    }
      
    /**
     * @return self
     */
    public function setFive($five) {
      $this->five = $five;
      return $this;
    }
  
    
    /**
     * @return 
     */
    public function getOne() {
      return $this->one;
    }
      
    /**
     * @return 
     */
    public function getTwo() {
      return $this->two;
    }
      
    /**
     * @return 
     */
    public function getThree() {
      return $this->three;
    }
      
    /**
     * @return 
     */
    public function getFour() {
      return $this->four;
    }
      
    /**
     * @return 
     */
    public function getFive() {
      return $this->five;
    }
  
}
"
`;
